Add the Data Validation help topic (#4647)
surveyjs/survey-library: docs/data-validation.md

| `"answercount"` | [`AnswerCountValidator`](https://surveyjs.io/Documentation/Library?id=AnswerCountValidator) | Throws an error if a user selects fewer choices than specified by [`minCount`](https://surveyjs.io/Documentation/Library?id=AnswerCountValidator#minCount) or more choices than specified by [`maxCount`](https://surveyjs.io/Documentation/Library?id=AnswerCountValidator#maxCount). Applies only to question types that can have multiple values (for instance, [Checkbox](https://surveyjs.io/Documentation/Library?id=questioncheckboxmodel)). |
| `"regex"` | [`RegexValidator`](https://surveyjs.io/Documentation/Library?id=RegexValidator) | Throws an error if an entered value does not match a regular expression defined in the [`regex`](https://surveyjs.io/Documentation/Library?id=RegexValidator#regex) property. |

- [View the "Built-In Validators" example](https://surveyjs.io/Examples/Library?id=validators-standard)
- [View the "Expression Validator" example](https://surveyjs.io/Examples/Library?id=validators-expression) 

## Implement Custom Client-Side Validation

> SurveyJS Library raises the [`onValidateQuestion`](https://surveyjs.io/Documentation/Library?id=surveymodel#onValidateQuestion) event that you can handle to add custom validation logic to your survey. For example, the following code checks that the answer to a `"memo"` question contains the word "survey":

Reviewer: ```suggestion
SurveyJS Library raises the [`onValidateQuestion`](https://surveyjs.io/Documentation/Library?id=surveymodel#onValidateQuestion) event, which you can handle to add custom validation logic to your survey. For example, the following code checks that the answer to a `"memo"` question contains the word "survey":
```

> SurveyJS Library raises the [`onValidateQuestion`](https://surveyjs.io/Documentation/Library?id=surveymodel#onValidateQuestion) event that you can handle to add custom validation logic to your survey. For example, the following code checks that the answer to a `"memo"` question contains the word "survey":

Reviewer 2: Subsection? ### Validate on Event

```js
import { Model } from "survey-core";
const surveyJson = {
  "elements": [{
    "name": "memo",
    "type": "comment",
    // ...
  }]
};
const survey = new Model(surveyJson);

survey.onValidateQuestion.add((survey, options) => {
  if (options.name === "memo") {

>     if (!!options.value.indexOf("survey")) {

Reviewer 2: May be it should be following condition
if (options.value.indexOf("survey") == -1) {

      options.error = 'Your answer must contain the word "survey"'
    }
  }
});
```

> Alternatively, you can implement custom validation via [expressions](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#expressions). Create a [custom function](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#custom-functions), register it, and then call it from within your expression. The following code uses this approach to demonstrate the use case above:

Reviewer 3: http://tw.devexpress.devx/language-guidelines/misused-words/via/

```suggestion
Alternatively, you can use [expressions](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#expressions) to implement custom validation. Create a [custom function](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#custom-functions), register it, and then call it from within your expression. The following code uses this technique to implement the same value validation scenario:
```

- `complete()` - A method that you should call when the request has completed. 

In the following example, a callback assigned to the `onServerValidateQuestions` event handler fetches a list of countries and checks if an entered country is in this list:

```js
import { Model } from "survey-core";
const surveyJson = {
  "elements": [{
    "name": "country",
    "type": "text",
    // ...
  }]
};
const survey = new Model(surveyJson);

function validateCountry(survey, { data, errors, complete }) {
  const countryName = data["country"];
  if (!countryName) {
    complete();
    return;
  }
  fetch("https://surveyjs.io/api/CountriesExample?name=" + countryName)
    .then((response) => {
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      return response.json();
    })
    .then((data) => {
      const found = data.length > 0;
      if (!found) {
        errors["country"] = "Country is not found";
      }
      complete();
    });
}

survey.onServerValidateQuestions.add(validateCountry);
```

[View example](https://surveyjs.io/Examples/Library?id=validators-server)

> Alternatively, you can implement custom validation via [expressions](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#expressions). Create an [asynchronous function](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#asynchronous-functions), register it, and then call it from within your expression. The following code uses this approach to demonstrate the use case above:

Reviewer 2: Subsection? ### Validation using `expression` built-in validator

Reviewer 2: Or ### Using custom functions in validation

> Alternatively, you can implement custom validation via [expressions](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#expressions). Create an [asynchronous function](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#asynchronous-functions), register it, and then call it from within your expression. The following code uses this approach to demonstrate the use case above:

Reviewer 3: http://tw.devexpress.devx/language-guidelines/misused-words/via/

```suggestion
Alternatively, you can use [expressions](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#expressions) to implement custom validation. Create an [asynchronous function](https://surveyjs.io/Documentation/Library?id=design-survey-conditional-logic#asynchronous-functions), register it, and then call it within your expression. The following code uses this technique to implement the previously demonstrated validation scenario:
```
